File: java-source/src/main/java/net/corda/training/HelloCorda.java
package net.corda.training;

import co.paralleluniverse.fibers.Suspendable;
import com.google.common.collect.ImmutableList;
import net.corda.core.contracts.*;
import net.corda.core.flows.*;
import net.corda.core.identity.AbstractParty;
import net.corda.core.identity.Party;
import net.corda.core.node.ServiceHub;
import net.corda.core.transactions.LedgerTransaction;
import net.corda.core.transactions.SignedTransaction;
import net.corda.core.transactions.TransactionBuilder;
import net.corda.core.utilities.ProgressTracker;
import org.jetbrains.annotations.NotNull;

import java.net.Proxy;
import java.security.SignatureException;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/* Flow */
public class HelloCorda {

    @InitiatingFlow
    static class SendMessageFlow extends FlowLogic<SignedTransaction> {
        private final Party target;
        public SendMessageFlow(Party target) { this.target = target; }

        @Suspendable
        @Override
        public SignedTransaction call() throws FlowException {
            // Get initiating party & notary
            final Party origin = getServiceHub().getMyInfo().getLegalIdentities().get(0);
            final Party notary = getServiceHub().getNetworkMapCache().getNotaryIdentities().get(0);

            // Create transaction items
            final MessageState state = new MessageState(origin, target, "Hello Corda!");
            final Command<MessageContract.SendMessage> command = new Command(new MessageContract.SendMessage(), Collections.singletonList(origin.getOwningKey()));

            // Create & sign transaction
            final StateAndContract stateAndContract = new StateAndContract(state, MessageContract.ID);
            TransactionBuilder builder = new TransactionBuilder(notary).withItems(stateAndContract, command);
            builder.verify(getServiceHub());

            /* TODO: EXCEPTION HAPPENS HERE */
            SignedTransaction stx = getServiceHub().signInitialTransaction(builder);

            // Send transaction to other party and finalize
            final FlowSession targetSession = initiateFlow(target);
            return subFlow(new FinalityFlow(stx, Arrays.asList(targetSession)));
        }
    }


    @InitiatedBy(SendMessageFlow.class)
    public static class SendMessageFlowResponder extends FlowLogic<SignedTransaction> {

        private final FlowSession counterpartySession;
        public SendMessageFlowResponder(FlowSession counterpartySession) {
            this.counterpartySession = counterpartySession;
        }

        @Suspendable
        @Override
        public SignedTransaction call() throws FlowException {
            return subFlow(new ReceiveFinalityFlow(counterpartySession));
        }

    }


}

/* Contract */
class MessageContract implements Contract {

    static final String ID = "net.corda.training.MessageContract";

    static class SendMessage extends TypeOnlyCommandData implements CommandData{}

    @Override
    public void verify(LedgerTransaction tx) {

        //final Commands commandData = requireSingleCommand(tx.getCommands(), Commands.class).getValue();
//        requireThat( require -> {
//            require.using("Must be a SendMessage command", commandData.equals(new Commands.SendMessage()));
//            require.using("There should be no input state.", tx.getInputStates().isEmpty());
//            require.using("There should one output state.", tx.getOutputStates().size() == 1);
//            require.using("The output state must be a MessageState.", tx.getOutputStates().get(0) instanceof MessageState);
//            return null;
//        });
    }
}



/* State */
@BelongsToContract(MessageContract.class)
class MessageState implements ContractState {
    final Party origin;
    final Party target;
    final String content;
    public MessageState(Party origin, Party target, String content) {
        this.origin = origin;
        this.target = target;
        this.content = content;
    }

    @NotNull
    @Override
    public List<AbstractParty> getParticipants() { return ImmutableList.of(origin, target); }
}

File: kotlin-source/src/main/kotlin/net/corda/training/HelloCorda.kt
package net.corda.training

import net.corda.core.contracts.*
import net.corda.core.flows.*
import net.corda.core.identity.Party
import net.corda.core.transactions.LedgerTransaction
import net.corda.core.transactions.SignedTransaction
import net.corda.core.transactions.TransactionBuilder
import net.corda.training.MessageContract.SendMessage
import co.paralleluniverse.fibers.Suspendable as Suspendable

/* Flow */
/* TODO(#1): Add an additional String parameter to SendMessageFlow so that we can pass a customized message.
    Then pass in this new parameter instead of "Hello World!" when we create our MessageState object. */
@InitiatingFlow
@StartableByRPC
class SendMessageFlow(private val target: Party) : FlowLogic<SignedTransaction>() {

    @Suspendable
    override fun call(): SignedTransaction {
        // Get initiating party & notary
        val origin = serviceHub.myInfo.legalIdentities.first()
        val notary = serviceHub.networkMapCache.notaryIdentities.single()

        // Create transaction items
        val command = Command(SendMessage(), listOf(origin.owningKey))
        val noteState = MessageState(origin, target, "") // ADD CONTENT
        val stateAndContract = StateAndContract(noteState, MessageContract.ID)

        // Create & sign transaction
        val builder = TransactionBuilder(notary).withItems(stateAndContract, command)
        val stx = serviceHub.signInitialTransaction(builder)
        stx.verify(serviceHub)

        // Send transaction to other party and finalize
        val targetSession = initiateFlow(target)
        return subFlow(FinalityFlow(stx, listOf(targetSession)))
    }
}
@InitiatedBy(SendMessageFlow::class)
class SendMessageFlowResponder(val counterpartySession: FlowSession) : FlowLogic<SignedTransaction>() {
    @Suspendable
    override fun call(): SignedTransaction {
        return subFlow(ReceiveFinalityFlow(counterpartySession))
    }
}

/* Contract */
/* TODO(#2): Add a constraint that ensures that our transaction has no input state! */
class MessageContract: Contract {
    companion object { const val ID = "net.corda.training.MessageContract" }

    // Commands
    class SendMessage : TypeOnlyCommandData()

    // Contract Clause
    override fun verify(tx: LedgerTransaction) = requireThat {
        val command = tx.commands.requireSingleCommand<SendMessage>()

        "There should be one output state." using (tx.outputStates.size == 1)
        val outputState = tx.outputStates.single() as MessageState
        "The party sending the message must sign the SendMessage transaction" using (command.signers == listOf(outputState.origin).map{ it.owningKey })

        // Add this
        // "There should be no input state." using tx.inputStates.isEmpty()
    }
}

/* State */
/* TODO(#3): Change the default message from "Hello World!" to "Hello Corda!" */
@BelongsToContract(MessageContract::class)
data class MessageState(val origin: Party, val target: Party, val content: String): ContractState {
    override val participants = listOf(origin, target)
}